okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.java

```diff
@@ -0,0 +1,142 @@
+ * request through the use of a URI Template.</blockquote>
+ */
+public class DnsOverHttps implements Dns {
+  public static final MediaType DNS_MESSAGE = MediaType.parse("application/dns-message");
+  public static final MediaType UDPWIREFORMAT = MediaType.parse("application/dns-udpwireformat");
+  private final OkHttpClient client;
+  private final HttpUrl url;
+  private final boolean includeIPv6;
+  private final boolean post;
+  private final MediaType contentType;
+
+  public DnsOverHttps(OkHttpClient client, HttpUrl url, @Nullable Dns bootstrapDns,
+      boolean includeIPv6, String method, MediaType contentType) {
+    this.client = bootstrapDns != null ? client.newBuilder().dns(bootstrapDns).build() : client;
+    this.url = url;
+    this.includeIPv6 = includeIPv6;
+    if (!method.equals("GET") && !method.equals("POST")) {
+      throw new UnsupportedOperationException("Only GET and POST Supported");
+    }
+    this.post = method.equals("POST");
+    this.contentType = contentType;
+  }
+
+  public HttpUrl getUrl() {
+    return url;
+  }
+
+  public boolean isPost() {
+    return post;
+  }
+
+  @Override public List<InetAddress> lookup(String hostname) throws UnknownHostException {
+    try {
+      ByteString query = DnsRecordCodec.encodeQuery(hostname, includeIPv6);
+
+      Request request = buildRequest(query);
+      Response response = executeRequest(request);
+
+      return readResponse(hostname, response);
+    } catch (UnknownHostException uhe) {
+      throw uhe;
+    } catch (Exception e) {
+      UnknownHostException unknownHostException = new UnknownHostException(hostname);
+      unknownHostException.initCause(e);
+      throw unknownHostException;
+    }
+  }
+
+  private Response executeRequest(Request request) throws IOException {
+    // cached request
+    if (!post && client.cache() != null) {
+      Request cacheRequest = request.newBuilder().cacheControl(CacheControl.FORCE_CACHE).build();
+
+      Response response = client.newCall(cacheRequest).execute();
+
+      if (response.isSuccessful()) {
+        return response;
+      }
+    }
+
+    return client.newCall(request).execute();
+  }
+
+  private List<InetAddress> readResponse(String hostname, Response response) throws Exception {
+    if (response.cacheResponse() == null && response.protocol() != Protocol.HTTP_2) {
+      Platform.get().log(Platform.WARN, "Incorrect protocol: " + response.protocol(), null);
+    }
+
+    try {
+      if (!response.isSuccessful()) {
+        throw new IOException("response: " + response.code() + " " + response.message());
+      }
+
+      ByteString responseBytes = response.body().source().readByteString();
+
+      List<InetAddress> results = DnsRecordCodec.decodeAnswers(hostname, responseBytes);
+
+      return results;
+    } finally {
+      response.close();
+    }
+  }
+
+  private Request buildRequest(ByteString query) {
+    Request.Builder builder;
```

Comment:
seems weird to have two different builder instances here

Reply:
... they both reference the same object!

Reply:
bad refactor!